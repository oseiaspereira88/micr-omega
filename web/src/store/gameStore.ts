import { useSyncExternalStore } from "react";
import {
  GamePhase,
  HealthState,
  Microorganism,
  OrganicMatter,
  Obstacle,
  OrientationState,
  RankingEntry,
  RoomObject,
  SharedGameState,
  SharedGameStateDiff,
  SharedPlayerState,
  SharedWorldState,
  Vector2,
} from "../utils/messageTypes";
import { reportRealtimeLatency } from "../utils/observability";

export type ConnectionStatus =
  | "idle"
  | "connecting"
  | "connected"
  | "reconnecting"
  | "disconnected";

export type PlayerMap = Record<string, SharedPlayerState>;

export type EntityCollection<T extends { id: string }> = {
  byId: Record<string, T>;
  all: T[];
};

type WorldCollections = {
  microorganisms: EntityCollection<Microorganism>;
  organicMatter: EntityCollection<OrganicMatter>;
  obstacles: EntityCollection<Obstacle>;
  roomObjects: EntityCollection<RoomObject>;
};

const createEmptyEntityCollection = <T extends { id: string }>(): EntityCollection<T> => ({
  byId: {},
  all: [],
});

const createEntityCollectionFromArray = <T extends { id: string }>(
  items: readonly T[],
  cloneItem: (item: T) => T,
): EntityCollection<T> => {
  const all = items.map((item) => cloneItem(item));
  const byId: Record<string, T> = {};
  for (const item of all) {
    byId[item.id] = item;
  }
  return { byId, all };
};

const applyEntityCollectionDiff = <T extends { id: string }>(
  collection: EntityCollection<T>,
  upserts: readonly T[] | undefined,
  removals: readonly string[] | undefined,
  cloneItem: (item: T) => T,
): { next: EntityCollection<T>; changed: boolean } => {
  let nextAll = collection.all;
  let nextById = collection.byId;
  let changed = false;

  if (upserts && upserts.length > 0) {
    nextAll = collection.all.slice();
    nextById = { ...collection.byId };
    const indexById = new Map<string, number>();
    nextAll.forEach((item, index) => {
      indexById.set(item.id, index);
    });

    for (const entry of upserts) {
      const cloned = cloneItem(entry);
      const existingIndex = indexById.get(cloned.id);
      if (existingIndex !== undefined) {
        nextAll[existingIndex] = cloned;
      } else {
        indexById.set(cloned.id, nextAll.push(cloned) - 1);
      }
      nextById[cloned.id] = cloned;
    }

    changed = true;
  }

  if (removals && removals.length > 0) {
    const removalSet = new Set(removals);
    if (removalSet.size > 0) {
      if (!changed) {
        nextAll = collection.all.slice();
        nextById = { ...collection.byId };
      }

      const filtered: T[] = [];
      let removedAny = false;
      for (const item of nextAll) {
        if (removalSet.has(item.id)) {
          delete nextById[item.id];
          removedAny = true;
        } else {
          filtered.push(item);
        }
      }

      if (removedAny) {
        nextAll = filtered;
        changed = true;
      }
    }
  }

  if (!changed) {
    return { next: collection, changed: false };
  }

  return {
    next: {
      byId: nextById,
      all: nextAll,
    },
    changed: true,
  };
};

const createEmptyWorldCollections = (): WorldCollections => ({
  microorganisms: createEmptyEntityCollection<Microorganism>(),
  organicMatter: createEmptyEntityCollection<OrganicMatter>(),
  obstacles: createEmptyEntityCollection<Obstacle>(),
  roomObjects: createEmptyEntityCollection<RoomObject>(),
});

const buildWorldFromCollections = (collections: WorldCollections): SharedWorldState => ({
  microorganisms: collections.microorganisms.all,
  organicMatter: collections.organicMatter.all,
  obstacles: collections.obstacles.all,
  roomObjects: collections.roomObjects.all,
});

export interface RoomStateSnapshot {
  phase: GamePhase;
  roundId: string | null;
  roundStartedAt: number | null;
  roundEndsAt: number | null;
}

export interface GameStoreState {
  connectionStatus: ConnectionStatus;
  reconnectAttempts: number;
  reconnectUntil: number | null;
  playerId: string | null;
  playerName: string | null;
  joinError: string | null;
  lastPingAt: number | null;
  lastPongAt: number | null;
  room: RoomStateSnapshot;
  players: PlayerMap;
  remotePlayers: EntityCollection<SharedPlayerState>;
  ranking: RankingEntry[];
  microorganisms: EntityCollection<Microorganism>;
  organicMatter: EntityCollection<OrganicMatter>;
  obstacles: EntityCollection<Obstacle>;
  roomObjects: EntityCollection<RoomObject>;
  world: SharedWorldState;
}

const defaultRoomState: RoomStateSnapshot = {
  phase: "waiting",
  roundId: null,
  roundStartedAt: null,
  roundEndsAt: null,
};

const createEmptySynchronizedState = () => {
  const remotePlayers = createEmptyEntityCollection<SharedPlayerState>();
  const worldCollections = createEmptyWorldCollections();
  return {
    remotePlayers,
    microorganisms: worldCollections.microorganisms,
    organicMatter: worldCollections.organicMatter,
    obstacles: worldCollections.obstacles,
    roomObjects: worldCollections.roomObjects,
    world: buildWorldFromCollections(worldCollections),
  };
};

const emptySyncState = createEmptySynchronizedState();

const initialState: GameStoreState = {
  connectionStatus: "idle",
  reconnectAttempts: 0,
  reconnectUntil: null,
  playerId: null,
  playerName: null,
  joinError: null,
  lastPingAt: null,
  lastPongAt: null,
  room: defaultRoomState,
  players: emptySyncState.remotePlayers.byId,
  remotePlayers: emptySyncState.remotePlayers,
  ranking: [],
  microorganisms: emptySyncState.microorganisms,
  organicMatter: emptySyncState.organicMatter,
  obstacles: emptySyncState.obstacles,
  roomObjects: emptySyncState.roomObjects,
  world: emptySyncState.world,
};

type GameStoreListener = () => void;

type StateUpdater = (state: GameStoreState) => GameStoreState;

type StateSelector<T> = (state: GameStoreState) => T;

let currentState: GameStoreState = initialState;
const listeners = new Set<GameStoreListener>();

const notify = () => {
  listeners.forEach((listener) => listener());
};

const applyState = (updater: StateUpdater) => {
  const nextState = updater(currentState);
  if (nextState !== currentState) {
    currentState = nextState;
    notify();
  }
};

const cloneVector = (vector: Vector2): Vector2 => ({ x: vector.x, y: vector.y });

const cloneOrientation = (orientation: OrientationState): OrientationState =>
  orientation.tilt === undefined
    ? { angle: orientation.angle }
    : { angle: orientation.angle, tilt: orientation.tilt };

const cloneHealth = (health: HealthState): HealthState => ({
  current: health.current,
  max: health.max,
});

const cloneCombatStatus = (
  status: SharedPlayerState["combatStatus"],
): SharedPlayerState["combatStatus"] => ({
  state: status.state,
  targetPlayerId: status.targetPlayerId,
  targetObjectId: status.targetObjectId,
  lastAttackAt: status.lastAttackAt,
});

const cloneCombatAttributes = (
  attributes: SharedPlayerState["combatAttributes"],
): SharedPlayerState["combatAttributes"] => ({
  attack: attributes.attack,
  defense: attributes.defense,
  speed: attributes.speed,
  range: attributes.range,
});

const clonePlayer = (player: SharedPlayerState): SharedPlayerState => ({
  ...player,
  position: cloneVector(player.position),
  movementVector: cloneVector(player.movementVector),
  orientation: cloneOrientation(player.orientation),
  health: cloneHealth(player.health),
  combatStatus: cloneCombatStatus(player.combatStatus),
  combatAttributes: cloneCombatAttributes(player.combatAttributes),
});

const cloneMicroorganism = (entity: Microorganism): Microorganism => ({
  ...entity,
  position: cloneVector(entity.position),
  movementVector: cloneVector(entity.movementVector),
  orientation: cloneOrientation(entity.orientation),
  health: cloneHealth(entity.health),
  attributes: { ...entity.attributes },
});

const cloneOrganicMatter = (matter: OrganicMatter): OrganicMatter => ({
  ...matter,
  position: cloneVector(matter.position),
  nutrients: { ...matter.nutrients },
});

const cloneObstacle = (obstacle: Obstacle): Obstacle => ({
  ...obstacle,
  position: cloneVector(obstacle.position),
  size: cloneVector(obstacle.size),
  orientation: obstacle.orientation ? cloneOrientation(obstacle.orientation) : undefined,
});

const cloneRoomObject = (object: RoomObject): RoomObject => ({
  ...object,
  position: cloneVector(object.position),
  state: object.state ? { ...object.state } : undefined,
});

const isFullState = (
  state: SharedGameState | SharedGameStateDiff
): state is SharedGameState => Array.isArray((state as SharedGameState).players);

const deriveRoomFromState = (
  state: SharedGameState | SharedGameStateDiff,
  previousRoom: RoomStateSnapshot
): RoomStateSnapshot => {
  if (isFullState(state)) {
    return {
      phase: state.phase,
      roundId: state.roundId,
      roundStartedAt: state.roundStartedAt,
      roundEndsAt: state.roundEndsAt,
    };
  }

  const nextRoom: RoomStateSnapshot = { ...previousRoom };

  if (state.phase !== undefined) {
    nextRoom.phase = state.phase;
  }

  if (state.roundId !== undefined) {
    nextRoom.roundId = state.roundId;
  }

  if (state.roundStartedAt !== undefined) {
    nextRoom.roundStartedAt = state.roundStartedAt;
  }

  if (state.roundEndsAt !== undefined) {
    nextRoom.roundEndsAt = state.roundEndsAt;
  }

  return nextRoom;
};

const setConnectionStatus = (status: ConnectionStatus) => {
  applyState((prev) => {
    if (prev.connectionStatus === status) {
      return prev;
    }

    const next: GameStoreState = {
      ...prev,
      connectionStatus: status,
    };

    if (status === "connected") {
      next.reconnectAttempts = 0;
    }

    if (status === "disconnected") {
      next.reconnectUntil = null;
    }

    return next;
  });
};

const setPlayerName = (name: string | null) => {
  applyState((prev) => {
    if (prev.playerName === name) {
      return prev;
    }

    return {
      ...prev,
      playerName: name,
    };
  });
};

const setPlayerId = (playerId: string | null) => {
  applyState((prev) => {
    if (prev.playerId === playerId) {
      return prev;
    }

    return {
      ...prev,
      playerId,
    };
  });
};

const setJoinError = (message: string | null) => {
  applyState((prev) => {
    if (prev.joinError === message) {
      return prev;
    }

    return {
      ...prev,
      joinError: message,
    };
  });
};

const incrementReconnectAttempts = () => {
  let attempts = 0;
  applyState((prev) => {
    attempts = prev.reconnectAttempts + 1;
    return {
      ...prev,
      reconnectAttempts: attempts,
    };
  });
  return attempts;
};

const setReconnectUntil = (timestamp: number | null) => {
  applyState((prev) => {
    if (prev.reconnectUntil === timestamp) {
      return prev;
    }

    return {
      ...prev,
      reconnectUntil: timestamp,
    };
  });
};

const applyFullState = (state: SharedGameState) => {
  applyState((prev) => {
    const remotePlayers = createEntityCollectionFromArray(state.players, clonePlayer);
    const worldCollections: WorldCollections = {
      microorganisms: createEntityCollectionFromArray(
        state.world.microorganisms,
        cloneMicroorganism,
      ),
      organicMatter: createEntityCollectionFromArray(state.world.organicMatter, cloneOrganicMatter),
      obstacles: createEntityCollectionFromArray(state.world.obstacles, cloneObstacle),
      roomObjects: createEntityCollectionFromArray(state.world.roomObjects, cloneRoomObject),
    };

    return {
      ...prev,
      room: {
        phase: state.phase,
        roundId: state.roundId,
        roundStartedAt: state.roundStartedAt,
        roundEndsAt: state.roundEndsAt,
      },
      players: remotePlayers.byId,
      remotePlayers,
      microorganisms: worldCollections.microorganisms,
      organicMatter: worldCollections.organicMatter,
      obstacles: worldCollections.obstacles,
      roomObjects: worldCollections.roomObjects,
      world: buildWorldFromCollections(worldCollections),
    };
  });
};

const applyStateDiff = (diff: SharedGameStateDiff) => {
  applyState((prev) => {
    const nextRoom = deriveRoomFromState(diff, prev.room);

    const playersResult = applyEntityCollectionDiff(
      prev.remotePlayers,
      diff.upsertPlayers,
      diff.removedPlayerIds,
      clonePlayer,
    );

    const worldDiff = diff.world;
    const microorganismsResult = applyEntityCollectionDiff(
      prev.microorganisms,
      worldDiff?.upsertMicroorganisms,
      worldDiff?.removeMicroorganismIds,
      cloneMicroorganism,
    );
    const organicMatterResult = applyEntityCollectionDiff(
      prev.organicMatter,
      worldDiff?.upsertOrganicMatter,
      worldDiff?.removeOrganicMatterIds,
      cloneOrganicMatter,
    );
    const obstacleResult = applyEntityCollectionDiff(
      prev.obstacles,
      worldDiff?.upsertObstacles,
      worldDiff?.removeObstacleIds,
      cloneObstacle,
    );
    const roomObjectResult = applyEntityCollectionDiff(
      prev.roomObjects,
      worldDiff?.upsertRoomObjects,
      worldDiff?.removeRoomObjectIds,
      cloneRoomObject,
    );

    const worldChanged =
      microorganismsResult.changed ||
      organicMatterResult.changed ||
      obstacleResult.changed ||
      roomObjectResult.changed;

    const nextWorld = worldChanged
      ? buildWorldFromCollections({
          microorganisms: microorganismsResult.next,
          organicMatter: organicMatterResult.next,
          obstacles: obstacleResult.next,
          roomObjects: roomObjectResult.next,
        })
      : prev.world;

    return {
      ...prev,
      room: nextRoom,
      players: playersResult.next.byId,
      remotePlayers: playersResult.next,
      microorganisms: microorganismsResult.next,
      organicMatter: organicMatterResult.next,
      obstacles: obstacleResult.next,
      roomObjects: roomObjectResult.next,
      world: nextWorld,
    };
  });
};

const applyRanking = (ranking: RankingEntry[]) => {
  applyState((prev) => ({
    ...prev,
    ranking,
  }));
};

const markPing = (timestamp: number) => {
  applyState((prev) => ({
    ...prev,
    lastPingAt: timestamp,
  }));
};

const markPong = (timestamp: number) => {
  applyState((prev) => {
    const latency =
      typeof prev.lastPingAt === "number" ? Math.max(0, timestamp - prev.lastPingAt) : null;
    if (latency !== null) {
      reportRealtimeLatency(latency, { source: "ws", phase: prev.room.phase });
    }

    return {
      ...prev,
      lastPongAt: timestamp,
    };
  });
};

const resetGameState = () => {
  const preservedName = currentState.playerName;
  const preservedId = currentState.playerId;
  const preservedStatus = currentState.connectionStatus;
  const preservedReconnectUntil = currentState.reconnectUntil;
  applyState(() => {
    const emptyState = createEmptySynchronizedState();
    return {
      ...initialState,
      playerName: preservedName,
      playerId: preservedId,
      connectionStatus: preservedStatus,
      reconnectUntil: preservedReconnectUntil,
      players: emptyState.remotePlayers.byId,
      remotePlayers: emptyState.remotePlayers,
      microorganisms: emptyState.microorganisms,
      organicMatter: emptyState.organicMatter,
      obstacles: emptyState.obstacles,
      roomObjects: emptyState.roomObjects,
      world: emptyState.world,
    };
  });
};

export const gameStore = {
  getState: () => currentState,
  subscribe: (listener: GameStoreListener) => {
    listeners.add(listener);
    return () => {
      listeners.delete(listener);
    };
  },
  setState: applyState,
  setPartial: (partial: Partial<GameStoreState>) => {
    applyState((prev) => ({
      ...prev,
      ...partial,
    }));
  },
  actions: {
    setConnectionStatus,
    setPlayerName,
    setPlayerId,
    setJoinError,
    incrementReconnectAttempts,
    setReconnectUntil,
    applyFullState,
    applyStateDiff,
    applyRanking,
    markPing,
    markPong,
    resetGameState,
  },
};

export function useGameStore<T>(selector: StateSelector<T>): T {
  return useSyncExternalStore(
    gameStore.subscribe,
    () => selector(currentState),
    () => selector(initialState)
  );
}
