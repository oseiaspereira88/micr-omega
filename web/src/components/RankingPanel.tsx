import { useMemo } from "react";
import { useGameStore } from "../store/gameStore";
import styles from "./RankingPanel.module.css";

const DISCONNECTED_ICON = "\u26A0\uFE0F"; // ⚠️

type RankingRow = {
  playerId: string;
  name: string;
  score: number;
  connected: boolean;
  isLocal: boolean;
};

const STATUS_LABEL: Record<string, string> = {
  idle: "Aguardando",
  connecting: "Conectando",
  connected: "Ao vivo",
  reconnecting: "Reconectando",
  disconnected: "Offline",
};

const STATUS_CLASS: Record<string, string> = {
  connected: styles.statusConnected,
  reconnecting: styles.statusReconnecting,
  disconnected: styles.statusDisconnected,
};

const RankingPanel = () => {
  const ranking = useGameStore((state) => state.ranking);
  const players = useGameStore((state) => state.players);
  const localPlayerId = useGameStore((state) => state.playerId);
  const connectionStatus = useGameStore((state) => state.connectionStatus);

  const rows = useMemo<RankingRow[]>(() => {
    return ranking.map((entry) => {
      const player = players[entry.playerId];
      return {
        playerId: entry.playerId,
        name: entry.name,
        score: entry.score,
        connected: player ? player.connected : false,
        isLocal: entry.playerId === localPlayerId,
      };
    });
  }, [localPlayerId, players, ranking]);

  if (rows.length === 0) {
    return (
      <aside className={styles.panel} aria-label="Ranking da partida">
        <header className={styles.header}>
          <h3 className={styles.title}>Ranking</h3>
          <span className={styles.statusBadge}>Sem dados</span>
        </header>
        <div className={styles.placeholder}>
          Aguarde o servidor enviar a classificação em tempo real.
        </div>
      </aside>
    );
  }

  const statusLabel = STATUS_LABEL[connectionStatus] ?? connectionStatus;
  const statusClass = STATUS_CLASS[connectionStatus] ?? "";

  return (
    <aside className={styles.panel} aria-label="Ranking da partida">
      <header className={styles.header}>
        <h3 className={styles.title}>Ranking</h3>
        <span className={`${styles.statusBadge} ${statusClass}`}>{statusLabel}</span>
      </header>
      <ol className={styles.list}>
        {rows.map((row, index) => {
          const tierClass =
            index === 0
              ? styles.top1
              : index === 1
              ? styles.top2
              : index === 2
              ? styles.top3
              : "";
          const itemClass = [styles.item, tierClass, row.isLocal ? styles.localPlayer : ""]
            .filter(Boolean)
            .join(" ");
          return (
            <li key={row.playerId} className={itemClass}>
              <span className={styles.rankBadge}>{index + 1}</span>
              <div className={styles.nameCell}>
                <span className={styles.name} title={row.name}>
                  {row.name}
                </span>
                {row.isLocal ? <span className={styles.localTag}>Você</span> : null}
              </div>
              <span className={styles.score}>
                {row.score.toLocaleString("pt-BR")}
                {!row.connected ? (
                  <span
                    className={styles.disconnected}
                    title="Jogador desconectado"
                    aria-label="Jogador desconectado"
                  >
                    {" "}
                    {DISCONNECTED_ICON}
                  </span>
                ) : null}
              </span>
            </li>
          );
        })}
      </ol>
    </aside>
  );
};

export default RankingPanel;
