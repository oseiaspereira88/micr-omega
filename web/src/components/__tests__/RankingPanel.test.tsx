import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { act, render, screen, within } from "@testing-library/react";
import RankingPanel from "../RankingPanel";
import { gameStore, type GameStoreState } from "../../store/gameStore";
import type { SharedPlayerState } from "../../utils/messageTypes";

const baseState = gameStore.getState();

const snapshot = (): GameStoreState => {
  const emptyPlayers = { byId: {}, all: [] };
  const emptyMicroorganisms = { byId: {}, all: [] };
  const emptyOrganic = { byId: {}, all: [] };
  const emptyObstacles = { byId: {}, all: [] };
  const emptyRoomObjects = { byId: {}, all: [] };

  return {
    ...baseState,
    room: { ...baseState.room },
    players: emptyPlayers.byId,
    remotePlayers: emptyPlayers,
    ranking: [],
    microorganisms: emptyMicroorganisms,
    organicMatter: emptyOrganic,
    obstacles: emptyObstacles,
    roomObjects: emptyRoomObjects,
    connectionStatus: "idle",
    playerId: null,
    playerName: null,
    joinError: null,
    world: {
      microorganisms: emptyMicroorganisms.all,
      organicMatter: emptyOrganic.all,
      obstacles: emptyObstacles.all,
      roomObjects: emptyRoomObjects.all,
      damagePopups: [],
    },
    damagePopups: [],
  };
};

const resetStore = () => {
  act(() => {
    gameStore.setState(() => snapshot());
  });
};

describe("RankingPanel", () => {
  beforeEach(() => {
    resetStore();
  });

  afterEach(() => {
    resetStore();
  });

  it("renders placeholder content when no ranking is available", () => {
    render(<RankingPanel />);

    expect(screen.getByRole("complementary", { name: "Ranking da partida" })).toBeInTheDocument();
    expect(screen.getByText("Aguardando")).toBeVisible();
    expect(
      screen.getByText("Aguardando o início da partida para receber o ranking.")
    ).toBeInTheDocument();
  });

  it("shows connection-aware placeholder messages", () => {
    render(<RankingPanel />);

    act(() => {
      gameStore.setState(() => ({
        ...snapshot(),
        connectionStatus: "disconnected",
      }));
    });

    expect(screen.getByText("Offline")).toBeVisible();
    expect(
      screen.getByText("Conexão perdida. Atualizaremos assim que o servidor responder novamente.")
    ).toBeInTheDocument();
  });

  it("highlights ordering, connection badges and the local player", () => {
    render(<RankingPanel />);

    act(() => {
      const players: GameStoreState["players"] = {
        alpha: createPlayer({
          id: "alpha",
          name: "Alice",
          connected: true,
          score: 4200,
          combo: 2,
          lastActiveAt: 10_000,
        }),
        beta: createPlayer({
          id: "beta",
          name: "Bruno",
          connected: false,
          score: 3150,
          combo: 1,
          lastActiveAt: 9_000,
        }),
      };
      const remotePlayers = {
        byId: players,
        all: Object.values(players),
      };

      gameStore.setState(() => ({
        ...snapshot(),
        connectionStatus: "connected",
        playerId: "beta",
        players,
        remotePlayers,
        ranking: [
          { playerId: "alpha", name: "Alice", score: 4200 },
          { playerId: "beta", name: "Bruno", score: 3150 },
        ],
      }));
    });

    expect(screen.getByText("Ranking")).toBeInTheDocument();
    expect(screen.getByText("Ao vivo")).toBeVisible();

    const items = screen.getAllByRole("listitem");
    expect(items).toHaveLength(2);

    const firstRow = within(items[0]!);
    expect(firstRow.getByText("1")).toBeVisible();
    expect(firstRow.getByText("Alice")).toBeVisible();
    expect(firstRow.getByText("4.200")).toBeVisible();
    expect(items[0]).toHaveAccessibleName(
      "1º lugar — Alice — 4.200 pontos — 100% da pontuação do líder — conectado",
    );

    const secondRow = within(items[1]!);
    expect(secondRow.getByText("2")).toBeVisible();
    expect(secondRow.getByText("Bruno")).toBeVisible();
    expect(secondRow.getByText("3.150")).toBeVisible();
    expect(secondRow.getByText("Você")).toBeVisible();
    expect(secondRow.getByLabelText("Jogador desconectado")).toBeVisible();
    expect(items[1]).toHaveAttribute("aria-current", "true");
    expect(items[1]).toHaveAccessibleName(
      "2º lugar — Você — 3.150 pontos — 75% da pontuação do líder — desconectado",
    );
  });

  it("matches snapshot for a populated ranking", () => {
    const { asFragment } = render(<RankingPanel />);

    act(() => {
      const players: GameStoreState["players"] = {
        alpha: createPlayer({
          id: "alpha",
          name: "Alice",
          connected: true,
          score: 4_200,
          combo: 2,
          lastActiveAt: 10_000,
        }),
        beta: createPlayer({
          id: "beta",
          name: "Bruno",
          connected: false,
          score: 3_150,
          combo: 1,
          lastActiveAt: 9_000,
        }),
      };
      const remotePlayers = {
        byId: players,
        all: Object.values(players),
      };

      gameStore.setState(() => ({
        ...snapshot(),
        connectionStatus: "connected",
        playerId: "beta",
        players,
        remotePlayers,
        ranking: [
          { playerId: "alpha", name: "Alice", score: 4_200 },
          { playerId: "beta", name: "Bruno", score: 3_150 },
        ],
      }));
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("orders ranking entries by score and name for ties", () => {
    render(<RankingPanel />);

    act(() => {
      const players: GameStoreState["players"] = {
        a: createPlayer({
          id: "a",
          name: "Ana",
          connected: true,
          score: 2_500,
          combo: 1,
          lastActiveAt: 1_000,
        }),
        b: createPlayer({
          id: "b",
          name: "Bruno",
          connected: true,
          score: 3_000,
          combo: 1,
          lastActiveAt: 2_000,
        }),
        c: createPlayer({
          id: "c",
          name: "Carla",
          connected: true,
          score: 3_000,
          combo: 1,
          lastActiveAt: 3_000,
        }),
      };
      const remotePlayers = {
        byId: players,
        all: Object.values(players),
      };

      gameStore.setState(() => ({
        ...snapshot(),
        connectionStatus: "connected",
        players,
        remotePlayers,
        ranking: [
          { playerId: "a", name: "Ana", score: 2_500 },
          { playerId: "c", name: "Carla", score: 3_000 },
          { playerId: "b", name: "Bruno", score: 3_000 },
        ],
      }));
    });

    const items = screen.getAllByRole("listitem");
    expect(items).toHaveLength(3);

    const [first, second, third] = items;
    expect(within(first!).getByText("Bruno")).toBeVisible();
    expect(within(second!).getByText("Carla")).toBeVisible();
    expect(within(third!).getByText("Ana")).toBeVisible();
  });

  it("ignores accents when comparing names and falls back to playerId", () => {
    render(<RankingPanel />);

    act(() => {
      const players: GameStoreState["players"] = {
        one: createPlayer({
          id: "one",
          name: "Ágata",
          connected: true,
          score: 4_000,
          combo: 1,
          lastActiveAt: 4_000,
        }),
        two: createPlayer({
          id: "two",
          name: "Agata",
          connected: true,
          score: 4_000,
          combo: 1,
          lastActiveAt: 5_000,
        }),
        three: createPlayer({
          id: "three",
          name: "Bruno",
          connected: true,
          score: 4_000,
          combo: 1,
          lastActiveAt: 6_000,
        }),
      };

      const remotePlayers = {
        byId: players,
        all: Object.values(players),
      };

      gameStore.setState(() => ({
        ...snapshot(),
        connectionStatus: "connected",
        players,
        remotePlayers,
        ranking: [
          { playerId: "three", name: "Bruno", score: 4_000 },
          { playerId: "two", name: "Agata", score: 4_000 },
          { playerId: "one", name: "Ágata", score: 4_000 },
        ],
      }));
    });

    const items = screen.getAllByRole("listitem");
    expect(items).toHaveLength(3);

    const [first, second, third] = items;
    expect(within(first!).getByText("Ágata")).toBeVisible();
    expect(within(second!).getByText("Agata")).toBeVisible();
    expect(within(third!).getByText("Bruno")).toBeVisible();
  });
});
const createPlayer = (
  overrides: Partial<SharedPlayerState> & Pick<SharedPlayerState, "id" | "name">
): SharedPlayerState => ({
  id: overrides.id,
  name: overrides.name,
  connected: overrides.connected ?? true,
  score: overrides.score ?? 0,
  combo: overrides.combo ?? 1,
  energy: overrides.energy ?? 0,
  xp: overrides.xp ?? 0,
  geneticMaterial: overrides.geneticMaterial ?? 0,
  lastActiveAt: overrides.lastActiveAt ?? 0,
  position: overrides.position ?? { x: 0, y: 0 },
  movementVector: overrides.movementVector ?? { x: 0, y: 0 },
  orientation: overrides.orientation ?? { angle: 0 },
  health: overrides.health ?? { current: 100, max: 100 },
  combatStatus:
    overrides.combatStatus ?? {
      state: "idle",
      targetPlayerId: null,
      targetObjectId: null,
      lastAttackAt: null,
    },
  combatAttributes:
    overrides.combatAttributes ?? {
      attack: 1,
      defense: 1,
      speed: 1,
      range: 1,
    },
  skillList: overrides.skillList ?? [],
  currentSkill: overrides.currentSkill ?? null,
  skillCooldowns: overrides.skillCooldowns ?? {},
});
