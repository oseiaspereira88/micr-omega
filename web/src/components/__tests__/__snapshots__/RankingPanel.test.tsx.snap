// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RankingPanel > matches snapshot for a populated ranking 1`] = `
<DocumentFragment>
  <aside
    aria-label="Ranking da partida"
    class="_panel_0f10a4"
  >
    <header
      class="_header_0f10a4"
    >
      <h3
        class="_title_0f10a4"
      >
        Ranking
      </h3>
      <span
        class="_statusBadge_0f10a4 _statusConnected_0f10a4"
      >
        Ao vivo
      </span>
    </header>
    <ol
      class="_list_0f10a4"
    >
      <li
        aria-label="1º lugar — Alice — 4.200 pontos — conectado"
        class="_item_0f10a4 _top1_0f10a4"
        role="row"
      >
        <span
          class="_rankBadge_0f10a4"
        >
          1
        </span>
        <div
          class="_nameCell_0f10a4"
        >
          <span
            class="_name_0f10a4"
            title="Alice"
          >
            Alice
          </span>
        </div>
        <span
          class="_score_0f10a4"
        >
          4.200
        </span>
      </li>
      <li
        aria-current="true"
        aria-label="2º lugar — Você — 3.150 pontos — desconectado"
        class="_item_0f10a4 _top2_0f10a4 _localPlayer_0f10a4"
        role="row"
      >
        <span
          class="_rankBadge_0f10a4"
        >
          2
        </span>
        <div
          class="_nameCell_0f10a4"
        >
          <span
            class="_name_0f10a4"
            title="Bruno"
          >
            Bruno
          </span>
          <span
            class="_localTag_0f10a4"
          >
            Você
          </span>
        </div>
        <span
          class="_score_0f10a4"
        >
          3.150
          <span
            aria-label="Jogador desconectado"
            class="_disconnected_0f10a4"
            title="Jogador desconectado"
          >
             ⚠️
          </span>
        </span>
      </li>
    </ol>
  </aside>
</DocumentFragment>
`;
